pandas/core/frame.py

```diff
@@ -3999,7 +4000,12 @@ def _combine_const(self, other, func, errors='raise', try_cast=True):
         new_data = self._data.eval(func=func, other=other,
                                    errors=errors,
                                    try_cast=try_cast)
-        return self._constructor(new_data)
+        result = self._constructor(new_data)
+        if is_number(other):
+            coerce_bool_cols = {col: type(other) for col in self.select_dtypes('bool')}
```

Comment:
maybe something more along the lines of 
```
In [18]: df = pd.DataFrame({'A': [True], 'B':[1], 'C':[False]})

In [19]: df
Out[19]: 
      A  B      C
0  True  1  False

# this is the internal routine, but appropriate here
In [20]: bt = df.dtypes.apply(pandas.core.dtypes.common.is_bool_dtype)

In [22]: i = bt[bt].index

In [23]: i
Out[23]: Index(['A', 'C'], dtype='object')
```

then
``result[i] = result[i].astype('bool')``

as select_dtypes copies things.